File: Python/6py 1/find_it.py
"""
|--------------------------------------------------------------------------
| Find the odd int
|--------------------------------------------------------------------------
|
"""

import functools
def find_it(seq):
   return functools.reduce(lambda a, b: a ^ b, seq)


def find_it(seq):
   return [x for x in seq if seq.count(x) % 2][0]


from itertools import groupby
def find_it(seq):
   return next(k for k, g in groupby(sorted(seq)) if len(list(g)) % 2)


def find_it(seq):
   odds = set()
   for i in seq:
      odds.symmetric_difference_update(set([i]))
   return odds.pop()


from collections import Counter
def find_it(seq):
   return [c for c in Counter(seq).items() if c[1] % 2 == 1][0][0]


print(find_it([20,1,-1,2,-2,3,3,5,5,1,2,4,20,4,-1,-2,5]), 5)
print(find_it([1,1,2,-2,5,2,4,4,-1,-2,5]), -1)
print(find_it([20,1,1,2,2,3,3,5,5,4,20,4,5]), 5)
print(find_it([10]), 10)
print(find_it([10, 10, 10]), 10)
print(find_it([1,1,1,1,1,1,10,1,1,1,1]), 10)
print(find_it([5,4,3,2,1,5,4,3,2,10,10]), 1)

File: PHP/6php 1/findIt.php
<?php
/*
|--------------------------------------------------------------------------
| Find the odd int
|--------------------------------------------------------------------------
|
*/

function findIt(array $seq) : int
{
   return array_reduce($seq, function($a, $b) {return $a ^ $b;});
}


function findIt(array $seq) : int
{
   foreach($seq as $val) if (array_count_values($seq)[$val] % 2) return $val;
}


function findIt(array $seq) : int {
   foreach (array_count_values($seq) as $i => $e) {
      if ($e % 2) {
         return $i;
      }
   }
}


function findIt(array $seq) : int {
   return key(array_filter(array_count_values($seq), function($a) {return $a % 2 !== 0;}));
}


print_r(findIt([20, 1, -1, 2, -2, 3, 3, 5, 5, 1, 2, 4, 20, 4, -1, -2, 5]). ' ' . 5 . '
');
print_r(findIt([1, 1, 2, -2, 5, 2, 4, 4, -1, -2, 5]). ' ' . -1 . '
');
print_r(findIt([20, 1, 1, 2, 2, 3, 3, 5, 5, 4, 20, 4, 5]). ' ' . 5 . '
');
print_r(findIt([10]). ' ' . 10 . '
');
print_r(findIt([1, 1, 1, 1, 1, 1, 10, 1, 1, 1, 1]). ' ' . 10 . '
');
print_r(findIt([5, 4, 3, 2, 1, 5, 4, 3, 2, 10, 10]). ' ' . 1 . '
');